arrow-libs/fx/arrow-collectors/src/commonMain/kotlin/arrow/collectors/Collector.kt

```diff
@@ -0,0 +1,237 @@
+     */
+    public fun <InternalAccumulator, Value, Result> nonSuspendOf(
+      supply: () -> InternalAccumulator,
+      accumulate: (current: InternalAccumulator, value: Value) -> Unit,
+      finish: (current: InternalAccumulator) -> Result,
+      characteristics: Set<Characteristics> = setOf(),
+    ): NonSuspendCollector<Value, Result> = object : NonSuspendCollectorI<InternalAccumulator, Value, Result> {
+      override val characteristics: Set<Characteristics> = characteristics
+      override fun supplyNonSuspend(): InternalAccumulator = supply()
+      override fun accumulateNonSuspend(current: InternalAccumulator, value: Value) {
+        accumulate(current, value)
+      }
+
+      override fun finishNonSuspend(current: InternalAccumulator): Result = finish(current)
+    }
+  }
+
+  /**
+   * Performs additional work during the finalization phase,
+   * by applying a function to the end result.
+   *
+   * @param transform Additional function to apply to the end result.
+   */
+  // functor over R
+  public fun <S> map(
+    transform: suspend (Result) -> S,
+  ): Collector<Value, S> = of(
+    supply = this::supply,
+    accumulate = this::accumulate,
+    finish = { current -> transform(this.finish(current)) },
+    characteristics = this.characteristics,
+  )
+
+  /**
+   * Applies a function over each element in the data source,
+   * before giving it to the collector.
+   *
+   * @param transform Function to apply to each value
+   */
+  // contravariant functor over T
+  public fun <P> contramap(
+    transform: suspend (P) -> Value,
+  ): Collector<P, Result> = of(
+    supply = this::supply,
+    accumulate = { current, value -> this.accumulate(current, transform(value)) },
+    finish = this::finish,
+    characteristics = this.characteristics,
+  )
+
+  /**
+   * Combines two [Collector]s by performing the phases
+   * of each of them in parallel.
+   *
+   * @param other [Collector] to combine with [this]
+   */
+  // applicative
+  public fun <B, S> zip(
+    other: CollectorI<B, @UnsafeVariance Value, S>,
+  ): Collector<Value, Pair<Result, S>> =
+    this.zip(other, ::Pair)
+
+  /**
+   * Combines two [Collector]s by performing the phases
+   * of each of them in parallel, and then combining
+   * the end result in a final step.
+   *
+   * @param other [Collector] to combine with [this]
+   * @param combine Function that combines the end results
+   */
+  public fun <B, S, V> zip(
+    other: CollectorI<B, @UnsafeVariance Value, S>,
+    combine: suspend (Result, S) -> V,
+  ): Collector<Value, V> = of(
+    supply = { Pair(this.supply(), other.supply()) },
+    accumulate = { (currentThis, currentOther), value ->
+      this.accumulate(currentThis, value)
+      other.accumulate(currentOther, value)
+    },
+    finish = { (currentThis, currentOther) ->
+      combine(this.finish(currentThis), other.finish(currentOther))
+    },
+    characteristics = this.characteristics,
```

Comment:
Hi @serras, I think the characteristics of the zipped collector should be the intersection of two input characteristics.
What do you think about that 🙏?

Reply:
I think that you're completely right!

Reply:
I've just pushed a fixed version. I also found that `map` was incorrect, since you were not guaranteed to have `IDENTITY_FINISH` in that case.